### PR TITLE
Add 2.x Scheduler to 1.x Scheduler conversion.

### PR DESCRIPTION
**Pull request by artem-zinnatullin**
Noticed a bug in our internal impl and decided to rewrite it and push to upstream.

@akarnokd do you think I should add v1 to v2 Scheduler conversion? We only need v2 to v1 since we're  aggressively replacing v1 with v2 and not vice versa.

### CODE REVIEW

File: src/test/java/hu/akarnokd/rxjava/interop/RxJavaInteropV2SchedulerToV1SchedulerTest.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+import org.junit.Test;
+import rx.functions.Action0;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.MINUTES;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.*;
+
+public class RxJavaInteropV2SchedulerToV1SchedulerTest {
```
Comment (artem-zinnatullin) [inline src/test/java/hu/akarnokd/rxjava/interop/RxJavaInteropV2SchedulerToV1SchedulerTest.java]: Decided to make separate test class, `RxJavaInteropTest` becomes kinda hard to maintain, hope you don't mind that :)

**Comment by artem-zinnatullin** [comment]
Not sure why codecoverage thinks I don't test that branch, I definitely do

screen shot 2018-01-26 at 2 03 19 pm

**Comment by akarnokd** [comment]
I thought about it before but didn't find any convincing reasoning for the conversion. You have the standard scheduler types for each version, so just pick the same type from the other one and you have `Schedulers.from` for the `Executor`s. I haven't encountered any custom scheduler implementations (beyond [my own ones](https://github.com/akarnokd/RxJava2Extensions#custom-schedulers) in the extensions project) out there.

**Comment by akarnokd** [comment]
> Not sure why codecoverage thinks I don't test that branch, I definitely do

You have an explicit `null` check which leads to the `null` check of `requireNonNull` that always throws, thus that particular line never completes normally and counts as not covered. Remove the `source == null` check and let `requireNonNull` handle it.

**Comment by artem-zinnatullin** [comment]
General idea is to only use minimal amount of underlying thread pools, rather than using two cpu cores bounded `computation()` from rx1 and rx2, as well as 2 unbounded `io()` pools.

So yeah, `Scheduler.fromExecutor()` is definitely an option to provide same backing thread pools for both rx1 and rx2, but tbh I prefer RxJava's Schedulers rather than Executors since they're "more native" for RxJava, already have expected configuration (ie bounded/unbounded) and in theory seem to have sliighly less overhead in compare to wrapped Executor into Scheduler, so we just use RxJava 2 schedulers and have this conversion for remaining rx1 to use same thread pools.

I'm totally fine with not merging it, so no pressure, but curios on what @vanniktech and @JakeWharton think about it

>You have an explicit null check which leads to the null check of requireNonNull that always throws, thus that particular line never completes normally and counts as not covered. Remove the source == null check and let requireNonNull handle it.

Facepalm… I've started with manual null check and exception, but then found that RxJavaInterop uses internal one from RxJava 2 and screwed up 😞

File: src/main/java/hu/akarnokd/rxjava/interop/SchedulerV2ToSchedulerV1.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
+package hu.akarnokd.rxjava.interop;
+
+
```
Comment (vanniktech) [inline src/main/java/hu/akarnokd/rxjava/interop/SchedulerV2ToSchedulerV1.java]: nit: double new line

File: src/main/java/hu/akarnokd/rxjava/interop/SchedulerV2ToSchedulerV1.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            public Subscription schedule(Action0 action, long delayTime, TimeUnit unit) {
+                return RxJavaInterop.toV1Subscription(
+                        v2Worker.schedule(new Action0V1ToRunnable(action), delayTime, unit)
+                );
+            }
+
+            @Override
+            public Subscription schedulePeriodically(Action0 action, long initialDelay, long period, TimeUnit unit) {
+                return RxJavaInterop.toV1Subscription(v2Worker.schedulePeriodically(
+                        new Action0V1ToRunnable(action), initialDelay, period, unit)
```
Comment (vanniktech) [inline src/main/java/hu/akarnokd/rxjava/interop/SchedulerV2ToSchedulerV1.java]: nit: instead of having it on two lines i'd just make variable out of it. it's also super hard to follow what exactly is v1 and what is v2

File: src/main/java/hu/akarnokd/rxjava/interop/SchedulerV2ToSchedulerV1.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    @Override
+    public long now() {
+        return source.now(MILLISECONDS);
+    }
+
+    @Override
+    public Worker createWorker() {
+        final Scheduler.Worker v2Worker = source.createWorker();
+
+        return new Worker() {
```
Comment (vanniktech) [inline src/main/java/hu/akarnokd/rxjava/interop/SchedulerV2ToSchedulerV1.java]: Let's also make this a static class?

File: src/test/java/hu/akarnokd/rxjava/interop/RxJavaInteropV2SchedulerToV1SchedulerTest.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+    @Test
+    public void now() {
+        Scheduler v2Scheduler = mock(Scheduler.class);
```
Comment (vanniktech) [inline src/test/java/hu/akarnokd/rxjava/interop/RxJavaInteropV2SchedulerToV1SchedulerTest.java]: Can't we use the normal `TestScheduler` here for testing? All of this white box mockito testing can be really brittle

File: src/test/java/hu/akarnokd/rxjava/interop/RxJavaInteropV2SchedulerToV1SchedulerTest.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        v1Scheduler.createWorker().schedule(action0, 123L, MINUTES);
+        verifyZeroInteractions(action0);
+
+        v2Scheduler.advanceTimeBy(122L, MINUTES);
+        verifyZeroInteractions(action0);
+
+        v2Scheduler.advanceTimeBy(1L, MINUTES);
+        verify(action0).call();
+
+        v2Scheduler.advanceTimeBy(125, MINUTES); // Check that it's not periodic.
```
Comment (vanniktech) [inline src/test/java/hu/akarnokd/rxjava/interop/RxJavaInteropV2SchedulerToV1SchedulerTest.java]: for consistency here's it missing the `L`

File: src/test/java/hu/akarnokd/rxjava/interop/RxJavaInteropV2SchedulerToV1SchedulerTest.java [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    @Test
+    public void workerUnsubscribe() {
+        Scheduler v2Scheduler = mock(Scheduler.class);
+        rx.Scheduler v1Scheduler = RxJavaInterop.toV1Scheduler(v2Scheduler);
+
+        Scheduler.Worker v2Worker = mock(Scheduler.Worker.class);
+        when(v2Scheduler.createWorker()).thenReturn(v2Worker);
+
+        rx.Scheduler.Worker v1Worker = v1Scheduler.createWorker();
+        verify(v2Worker, times(0)).dispose();
```
Comment (vanniktech) [inline src/test/java/hu/akarnokd/rxjava/interop/RxJavaInteropV2SchedulerToV1SchedulerTest.java]: nit: we can use `never()` instead of `times(0)`

File: src/test/java/hu/akarnokd/rxjava/interop/RxJavaInteropV2SchedulerToV1SchedulerTest.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        Scheduler v2Scheduler = mock(Scheduler.class);
+        rx.Scheduler v1Scheduler = RxJavaInterop.toV1Scheduler(v2Scheduler);
+
+        Scheduler.Worker v2Worker = mock(Scheduler.Worker.class);
+        when(v2Scheduler.createWorker()).thenReturn(v2Worker);
+
+        rx.Scheduler.Worker v1Worker = v1Scheduler.createWorker();
+
+        when(v2Worker.isDisposed()).thenReturn(true);
+        assertEquals(true, v1Worker.isUnsubscribed());
```
Comment (vanniktech) [inline src/test/java/hu/akarnokd/rxjava/interop/RxJavaInteropV2SchedulerToV1SchedulerTest.java]: nit: assertTrue?

File: src/test/java/hu/akarnokd/rxjava/interop/RxJavaInteropV2SchedulerToV1SchedulerTest.java [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+
+        when(v2Worker.isDisposed()).thenReturn(false);
+        assertEquals(false, v1Worker.isUnsubscribed());
```
Comment (vanniktech) [inline src/test/java/hu/akarnokd/rxjava/interop/RxJavaInteropV2SchedulerToV1SchedulerTest.java]: nit: assertFalse

**Comment by JakeWharton** [comment]
Having double the thread pools is motivation to complete the transition sooner!

We decided to just use both scheduler versions and hope it didn't have any adverse effects. I'm not against the converter, though. It'd certainly enable a bulk change with IntelliJ's structured replace and then I would never think of it again.

File: src/test/java/hu/akarnokd/rxjava/interop/RxJavaInteropV2SchedulerToV1SchedulerTest.java (hunk 5 above)
Comment (JakeWharton) [reply src/test/java/hu/akarnokd/rxjava/interop/RxJavaInteropV2SchedulerToV1SchedulerTest.java]: A comment to remove Mockito usage? Instant Jake-thumbs-up to that: 👍

Comment (artem-zinnatullin) [reply src/test/java/hu/akarnokd/rxjava/interop/RxJavaInteropV2SchedulerToV1SchedulerTest.java]: Also not a fan of Mockito and mocking, however in scenarios like proxying which is what I do here it's saves us from boilerplate and keeps code readable

Especially in few tests below when I need to verify invocations

**Comment by artem-zinnatullin** [comment]
>Having double the thread pools is motivation to complete the transition sooner!

Unfortunately for many projects that means months of degraded user experience (higher memory usage, worse threads swiching)

>It'd certainly enable a bulk change with IntelliJ's structured replace and then I would never think of it again.

That's very questionable in practice, specifically due to dropped support of `null`s as `onNext` payload, I'm part of 2nd big project migration (Juno and Lyft) and have to say that daaaaaamn it's complicated sometimes 

Juno's approach was discussed by @ming13 and @sockeqwe in Episode 11 of The Context podcast: ["Migration to RxJava 2 with Artur Dryomov from Juno"](https://github.com/artem-zinnatullin/TheContext-Podcast/releases/tag/Episode_11)

Took about 4 months of gradual migration in ~50k Kotlin codebase where ~everything was reactive.

At Lyft we're still doing migration for about half a year now in a 500k LOC Java codebase, so it's definitely not a simple structural replace unfortunately 😿

**Comment by JakeWharton** [comment]
Degraded user experience is a stretch. And you seem to have misinterpreted my comment as applying to all of RxJava instead of just the schedulers.

**Comment by artem-zinnatullin** [comment]
Ahh, now it makes much more sense 👍

**Comment by JakeWharton** [comment]
To be clear: I'm saying I would use this contribution and find/replace all
the uses of 1.x schedulers to be 2.x ones backported and this change makes
it easy to do so. So I'm supportive.

**Comment by artem-zinnatullin** [comment]
I'd probably recommend to apply it in one place instead of structural replace, ie through `RxJavaHooks.setOn*Scheduler` like that:

```java
RxJavaHooks.setOnComputationScheduler(scheduler -> {
    io.reactivex.Scheduler computation = Schedulers.computation();
    return new RxJava1To2Scheduler(computation);
});
```

Or one could do it in a DI module, Juno injects schedulers for instance

---

>To be clear

I would pay for an OS extension that'll do this with all your messages in all apps

**Comment by ming13** [comment]
The idea is great, not sure why we didn’t do it internally in the first place.

BTW I don’t remember that the `Rx*` threads count was that much higher than a baseline during the migration process. Maybe that was so because most of the heavy-lifting was migrated last. Or we just don’t have a Lyft _scale_ 😉

File: src/main/java/hu/akarnokd/rxjava/interop/RxJavaInterop.java [hunk 10]
```diff
@@ -16,6 +16,8 @@
 
 package hu.akarnokd.rxjava.interop;
 
+import io.reactivex.Scheduler;
```
Comment (akarnokd) [inline src/main/java/hu/akarnokd/rxjava/interop/RxJavaInterop.java]: Please don't use imports of Rx stuff in this file but use the fully qualified name instead, like the other methods.

File: src/main/java/hu/akarnokd/rxjava/interop/SchedulerV2ToSchedulerV1.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+import io.reactivex.Scheduler;
+import rx.Subscription;
+import rx.functions.Action0;
+
+import java.util.concurrent.TimeUnit;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+final class SchedulerV2ToSchedulerV1 extends rx.Scheduler {
```
Comment (akarnokd) [inline src/main/java/hu/akarnokd/rxjava/interop/SchedulerV2ToSchedulerV1.java]: Please include a short javadoc and the `@since 0.12.0` annotation here too.

**Comment by akarnokd** [comment]
Would you also implement the v1-to-v2 conversion as well, for symmetry (also there is a high chance people surely will ask about why that is missing)? It can be in a separate PR.

**Comment by artem-zinnatullin** [comment]
Ok, will address your comments in separate PR(s), thanks!

**Comment by akarnokd** [comment]
Don't, I'm currently making the changes and some other additions (such as the shutdown support).
